compiler/src/iree/compiler/Utils/Indexing.h

```diff
@@ -0,0 +1,78 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_UTILS_SHAPE_H_
+#define IREE_COMPILER_UTILS_SHAPE_H_
+
+#include <algorithm>
+#include <iterator>
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Support/LLVM.h"
+
+namespace mlir::iree_compiler {
+
+// Make a list of expressions that computes the stride for each dimension from a
+// shape.
+inline SmallVector<AffineExpr> stridesFromShape(ArrayRef<AffineExpr> shape) {
+  SmallVector<AffineExpr> strides;
+  if (shape.empty()) {
+    return strides;
+  }
+
+  AffineExpr runningProduct = getAffineConstantExpr(1, shape[0].getContext());
+  SmallVector<AffineExpr> reverseStrides;
+  std::transform(shape.rbegin(), shape.rend(),
+                 std::back_inserter(reverseStrides),
+                 [&runningProduct](AffineExpr shapeElement) {
+                   AffineExpr res = runningProduct;
+                   runningProduct = shapeElement * runningProduct;
+                   return res;
+                 });
+  std::reverse_copy(reverseStrides.begin(), reverseStrides.end(),
+                    std::back_inserter(strides));
+
+  return strides;
+}
+
+// Construct IR that extracts the linear index form a multi-index according to
+// a shape.
+inline Value linearIndex(ArrayRef<Value> multiIndex, ArrayRef<Value> shape,
```

Comment:
This is sort of duplicating `mlir::computeLinearIndex` in https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/IndexingUtils.h? Can we use that one?

Reply:
I removed `linearIndex` and made the other code use `mlir::computeLinearIndex`.